scripts/build/path-resolver.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Parses and resolves filesystem paths.
+ */
+
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+module.exports = class {
+  constructor({
+    pathLib = path,
+    fsLib = fs,
+  } = {}) {
```

Comment:
Great question!

Object params have some really nice properties (no pun intended):

1. They're not order-dependent, so:
    1. Callers can pass them in any order
    2. It's impossible to accidentally pass params in the wrong order \
        (I've seen lots of bugs caused by that)
    3. We can reorder the params however we want, without breaking callers. Positional params, on the other hand, force you to put ALL params with default values at the very end.
    4. If we use descriptive param names (which I try to), it forces callers to explicitly declare the name of each param, which can make certain types of data mismatch easier to see. \
        E.g.:
        ```js
        // Are these params in the right order? Who knows!
        foo('**/*.scss', '[name].css')
        ```
        vs.
        ```js
        foo({
          inputFilePathPattern: '**/*.scss',
          outputFilenamePattern: '[name.css]',
        })
        ```
2. They act as documentation for the expected structure of nested config objects \
    (e.g., [`createCustomCss()` in `bundle-factory-js`](https://github.com/material-components/material-components-web/blob/355d2c5fa4b90687e55b9cf48624c3d7db675eba/scripts/webpack/css-bundle-factory.js#L49))

The default empty object is needed to allow the user to avoid passing any params at all. If you try to reference one of the params inside a function, and the user didn't pass _anything_, you'll get an error like this:

```
scripts/build/path-resolver.js:28
    pathLib = path,
    ^

TypeError: Cannot destructure property `pathLib` of 'undefined' or 'null'.
    at new module.exports (scripts/build/path-resolver.js:27:14)
    at Object.<anonymous> (webpack.config.js:30:22)
```

Reply:
+1 to param objects in cases where most or all of the parameters are optional. Way less ugly than positionally-optional arguments, for sure...

Question in this particular case, looking at it out of context though: Why would we ever have a reason to use anything besides the stock `fs` and `path` modules?

Reply:
That's only for unit testing, so we can mock the libraries to return whatever we want.

Sample mock:
![image](https://user-images.githubusercontent.com/409245/36123039-4b0f4778-1000-11e8-9daf-38c056233d00.png)